Parse layer index from G-code comments

saumyaj3 opened the pull request:
This update includes logic to parse and cache the layer index from layer comments found in G-code files. This change ensures that layer start and end points are properly identified and stored, improving the accuracy of layer-based operations such as 3D printing visualizations.

- Added an extra option from the user perspective to use comments from gcode to determine layer.
- Corrected the radius of the extrusion as it was increasing in squared way before.

saumyaj3 committed:
Parse layer index from G-code comments

This update includes logic to parse and cache the layer index from layer comments found in G-code files. This change ensures that layer start and end points are properly identified and stored, improving the accuracy of layer-based operations such as 3D printing visualizations.

NP-639

aligator commented on src/parser.ts:
```diff
@@ -231,103 +231,100 @@ export class GCodeParser {
 
-                // Split off comments.
-                line = line.split(";", 2)[0]
+                if (line.includes(";LAYER:")) {
```
Maybe it is better to use a regexp for this?
You could first check for the ';' (no need to check every line with the regexp) and then execute a regexp that returns the number.
e.g. /^;LAYER:(\d+)/

saumyaj3 replied:
I implemented this here [f14892a](https://github.com/aligator/gcode-viewer/pull/26/commits/f14892a9ef812022434231608a23fa6db5be7160)

aligator reviewed (changes requested):
Hi, 
Thanks for the PR,

I did not test it yet, but here are already some hints for improvement:

* Note to myself - I haven't set up / defined a code formatter yet, a bit hard to review.
-> I pushed everything formatted with prettier (with the default prettier settings) to the main branch. Could you update your branch and also format with prettier? Would be much more easy to review.
* You removed several comments, I think most of them are still correct, could you please preserve them (or update them if they are no longer correct)
* If I understand it correctly, you removed the old way of determining the Index with the new way.  
-> since not all gcodes may have the line number set as comment, I think it would be better to add a configuration for this. (maybe even with a configurable Regexp?)
(However i think since nowadays all slicers should add this comment, it is ok to use this behavior as default? A bit of a breaking change though...)

maybe some option like this:
```ts
  /**
   * Regexp to determine the layer from the gcode comments.
   *
   * If set to a regexp such as `/^;LAYER:(\d+)/` the layer will be determined by searching for this regexp.
   * If it is undefined, the layer will be determined by trying to find z-value changes which may not be as accurate.
   *
   * @default /^;LAYER:(\d+)/
   * @type RegExp | undefined
   */
  public layerRegexp?: RegExp = /^;LAYER:(\d+)/;
```

saumyaj3 commented:
hi @aligator! Thanks for a quick review. I think the solution I have here is pretty specific to curaengine produced gcode and might not work for general slicer. I wanted to make a fork and then have the changes. I think I did a bit of mistake. I will delete this PR. However, Thankyou for the library. :)

saumyaj3 committed:
Merge branch 'main' into NP-639-create-correct-layer-number

# Conflicts:
#	src/parser.ts

saumyaj3 committed:
Indent HTML and JavaScript for better readability

NP-639

saumyaj3 committed:
Refactor layer index parsing logic

Update the logic for parsing layer indices by using regex and checking lines that start with a semicolon. This change ensures more robust and accurate identification of layer information in the parser.

NP-639

saumyaj3 committed:
Refactor parser.ts with clearer comment annotations

NP-639

saumyaj3 committed:
Add comment to clarify getValue function usage

NP-639

aligator commented:
@saumyaj3 
ok, Nice to see it used somewhere :-) 

I will observe the changes in the fork, maybe I can make use of some things there in the future and port them back to my lib :+1:

saumyaj3 committed:
Add LayerType enum and logic to handle layer slicing

Introduce LayerType enum to specify layer slicing by VARIABLE_Z or LAYER_COMMENTS. Updated parser and README to handle these configurations, allowing customization based on z-values or GCode comments for layer segmentation.

NP-639

saumyaj3 committed:
Put the previous example back

saumyaj3 commented:
@aligator I reopened it. made the suggested change and also, made it a bit generic, so users can choose how they want it to be displayed. I mean the layerview. It is better if the changes can reflect in the library itself ;)

saumyaj3 committed:
Calculate radius using square root of squared value

This change refactors the calculation of the `radius` variable by introducing a `radiusSquared` intermediate step. The code now correctly computes the radius by taking the square root, which should result in more accurate calculations. Negative extrusions are still hidden as intended.

NP-639

saumyaj3 committed:
Use const for radiusSquared variable

aligator commented on README.md:
```diff
@@ -101,6 +101,15 @@ renderer.travelWidth = 0.1;
+### Gcode LayerViews
+
+The default for sectioning the layers is determining the layers based on changing z-values on the gcode.
+However, If the user wants layers to be taken from the comments of the Gcode. they can specify the lineType as:
```
```suggestion
However, If the user wants layers to be taken from the comments of the Gcode such as `;LAYER:1`. they can specify the lineType as:
```

aligator commented on src/gcode.ts:
```diff
@@ -71,6 +71,21 @@ export class GCodeRenderer {
   }
 
+  /**
+   * Type to determine how the layers are sliced
```
```suggestion
   * Type to determine how the layer change is detected.
```

aligator commented on src/gcode.ts:
```diff
@@ -71,6 +71,21 @@ export class GCodeRenderer {
+   }
+
+  /**
+   * Type to determine how the layers are sliced
```
```suggestion
   * Type to determine how the layer change is detected.
```

aligator commented on src/parser.ts:
```diff
@@ -80,6 +80,11 @@ export interface LayerDefinition {
 }
 
+export enum LayerType{
+  VARIABLE_Z,
+  LAYER_COMMENTS,
```
```suggestion
  /**
   * Layers are defined by the Z value of the points.
   * This may not be accurate, and may have problems if used with Z-Hops or similar things.
   */
  VARIABLE_Z,

  /**
   * Layer changes are defined by Layer comments in the GCode.
   * This is more accurate but requires the GCode to have layer comments.
   * ;LAYER:0
   */
  LAYER_COMMENTS,
```

aligator commented on src/parser.ts:
```diff
@@ -131,6 +136,8 @@ export class GCodeParser {
    */
   public radialSegments: number = 8;
 
+  public layerType: LayerType = LayerType.VARIABLE_Z;
```
```suggestion
  /**
   * The layer type to determine how the layer change is detected.
   * @type LayerType
   * @default LayerType.VARIABLE_Z
   */
  public layerType: LayerType = LayerType.VARIABLE_Z;
```

aligator reviewed (changes requested):
LGTM, Just some more comment suggestions.
Thanks for your contribution :-)

saumyaj3 committed:
Apply suggestions from code review

Co-authored-by: aligator <johannes.hoermann@t-online.de>

saumyaj3 committed:
Merge branch 'main' into NP-639-create-correct-layer-number

saumyaj3 commented:
@aligator  thanks for such quick review. :dancers: . I just wanted to ask if you're planning to have a package released soon?

aligator commented:
When you are done with all things you want to improve, just tell me and I will release a new version.

saumyaj3 commented:
Hey @aligator That would be really nice :D with this PR and [PR-28](https://github.com/aligator/gcode-viewer/pull/28) . Thankyou :dancing_women: 

I will have some more features coming up soon probably. I'll keep them generic. :100:

aligator commented:
> Hey @aligator That would be really nice :D with this PR and [PR-28](https://github.com/aligator/gcode-viewer/pull/28) . Thankyou 👯‍♀️
> 
> I will have some more features coming up soon probably. I'll keep them generic. 💯

Should be released now as 0.6.2
